L1Trigger/L1THGCal/src/be_algorithms/HGCalHardwareInterface.cc

```diff
@@ -0,0 +1,205 @@
+		MultiClusterDataMap_[MultiClusterDataFormat::MaxEnergyLayer]);  
+  return true;
+}
+
+void HGCalHardwareInterface::InitDataMap_(){
+  /* Data format bit-size for each word */  
+  ClusterDataMap_[ClusterDataFormat::ET]            = 8;
+  ClusterDataMap_[ClusterDataFormat::CenterY]       = 12;
+  ClusterDataMap_[ClusterDataFormat::CenterX]       = 12;
+  ClusterDataMap_[ClusterDataFormat::Flags]         = 6;
+  ClusterDataMap_[ClusterDataFormat::SizeY]         = 8;
+  ClusterDataMap_[ClusterDataFormat::SizeX]         = 8;
+  ClusterDataMap_[ClusterDataFormat::LocalMax]      = 2;
+  ClusterDataMap_[ClusterDataFormat::NoTriggerCells]= 8;
+  ClusterDataMap_[ClusterDataFormat::LocalMax0RelY] = 8;  
+  ClusterDataMap_[ClusterDataFormat::LocalMax0RelX] = 8;
+  ClusterDataMap_[ClusterDataFormat::LocalMax0ET]   = 8;  
+  ClusterDataMap_[ClusterDataFormat::LocalMax1RelY] = 8;  
+  ClusterDataMap_[ClusterDataFormat::LocalMax1RelX] = 8;
+  ClusterDataMap_[ClusterDataFormat::LocalMax1ET]   = 8;  
+  ClusterDataMap_[ClusterDataFormat::LocalMax2RelY] = 8;  
+  ClusterDataMap_[ClusterDataFormat::LocalMax2RelX] = 8;
+  ClusterDataMap_[ClusterDataFormat::LocalMax2ET]   = 8;  
+  ClusterDataMap_[ClusterDataFormat::LocalMax3RelY] = 8;  
+  ClusterDataMap_[ClusterDataFormat::LocalMax3RelX] = 8;
+  ClusterDataMap_[ClusterDataFormat::LocalMax3ET]   = 8;  
+
+  MultiClusterDataMap_[MultiClusterDataFormat::BHEFraction]    = 8;
+  MultiClusterDataMap_[MultiClusterDataFormat::EEEFraction]    = 8;
+  MultiClusterDataMap_[MultiClusterDataFormat::ET]             = 16;
+  MultiClusterDataMap_[MultiClusterDataFormat::Z]              = 10;
+  MultiClusterDataMap_[MultiClusterDataFormat::Phi]            = 11;
+  MultiClusterDataMap_[MultiClusterDataFormat::Eta]            = 11;
+  MultiClusterDataMap_[MultiClusterDataFormat::Flags]          = 12;
+  MultiClusterDataMap_[MultiClusterDataFormat::QFlags]         = 12;
+  MultiClusterDataMap_[MultiClusterDataFormat::NoTriggerCells] = 8;
+  MultiClusterDataMap_[MultiClusterDataFormat::TBA]            = 26;
+  MultiClusterDataMap_[MultiClusterDataFormat::MaxEnergyLayer] = 6;
+
+}
+
+short  HGCalHardwareInterface::get32bitPart_(const std::uint32_t data32bit, const unsigned short lsb, const unsigned short size )
+{
+  return (data32bit >> lsb) & (0xffffffff >> (32-size));
+}
+
+void  HGCalHardwareInterface::del32bitPart_(std::uint32_t & data32bit,  const unsigned short lsb, const unsigned short size )
+{
+  data32bit = data32bit &  (((lsb > 0) ? (0xffffffff  >> (32-lsb)) : 0x0 ) | ((lsb+size > 31)? 0x0 : 0xffffffff  << (lsb+size)));
+}
+
+void  HGCalHardwareInterface::set32bitPart_(std::uint32_t & data32bit, const std::int32_t bitvalue,  const unsigned short lsb, const unsigned short size )
+{
+  if( bitvalue > std::pow(2, size) - 1 || -1 * bitvalue > std::pow(2,size-1) )
+    throw std::out_of_range("the value exceeds the given bit range");
+  if(size > 32 || lsb > 31 )
+    throw std::out_of_range("the value exceeds 32-bit");
```

Comment:
CMSSW handles better `cms::Exception`. The goal is to provide better information concerning the source of the problem.